src/config.rs

```diff
@@ -74,10 +75,235 @@ pub fn lookup(dir: &Path) -> Option<PathBuf> {
+        "#;
+        let fmt_config = read_toml::<FmtConfig>(toml).unwrap();
+        let default_excludes: Vec<String> = Vec::new();
+        assert_eq!(default_excludes, fmt_config.options);
+
+        // parses correctly
+        let spec = r#"
+        command="sh"
+        excludes=["foo", "bar", "baz"]
+        "#;
+        let fmt_config = read_toml::<FmtConfig>(spec).unwrap();
+        assert_eq!(vec!["foo", "bar", "baz"], fmt_config.excludes);
+    }
+
+    #[test]
+    fn test_root() {
+        // requires at least one formatter
+        let toml = "";
+        assert_eq!(
+            Err("missing field `formatter`".to_string()),
+            read_toml::<Root>(toml)
+        );
+
+        // global defaults to None
+        let toml = r#"
+        [formatter.sh]
+        command = "sh"
+        "#;
+
+        let expected = default_root("sh");
+        assert_eq!(Ok(expected), read_root(toml));
+
+        // parses global correctly
+        let toml = r#"
+        [global]
+        excludes = ["foo", "bar", "baz"]
+
+        [formatter.sh]
+        command = "sh"
+        "#;
+        let mut expected = default_root("sh");
+        expected.global = Some(GlobalConfig {
+            excludes: ["foo", "bar", "baz"].map(String::from).to_vec(),
+        });
+
+        assert_eq!(Ok(expected), read_root(toml));
+
+        // formatters are keyed correctly in the formatters map
+        let toml = r#"
+        [formatter.sh]
+        command = "sh"
+
+        [formatter.rust]
+        command = "rustfmt"
+
+        [formatter.haskell]
+        command = "cabal-fmt"
+        "#;
+
+        let mut expected = default_root("sh");
+        expected.formatter = BTreeMap::from([
+            ("sh".to_string(), default_fmt_config("sh")),
+            ("rust".to_string(), default_fmt_config("rustfmt")),
+            ("haskell".to_string(), default_fmt_config("cabal-fmt")),
+        ]);
+
+        assert_eq!(Ok(expected), read_root(toml));
+    }
+
+    #[test]
+    fn test_from_path() {
+        // utility for mapping into Result<Root, String> for easier assertions
+        let read_from_path = |path: &Path| -> Result<Root, String> {
+            from_path(path).map_err(|x| x.to_string())
+        };
+
+        // bad path
+        assert_eq!(Err("No such file or directory (os error 2)".to_string()), read_from_path(Path::new("foo/bar/baz.toml")));
+
+        // good path
+        let mut expected = default_root("sh");
+        expected.global = Some(GlobalConfig {
+            excludes: ["foo", "bar", "baz"].map(String::from).to_vec(),
+        });
+
+        // the cwd directory during `cargo test` is the same directory as the crate under test
+        // we do the following to ensure a consistent path regardless of running in IDE or terminal
+        let path = concat!(env!("CARGO_MANIFEST_DIR"), "/examples/treefmt.toml");
+        let path = PathBuf::from(path).canonicalize().unwrap();
+
+        assert_eq!(Ok(expected), read_from_path(&path));
```

Comment:
@zimbatm this test passes fine during `cargo test` and when running it manually in my IDE with arbitrary working directories. 

However it fails when you run `nix fmt`, e.g. build the `treefmt` nixpkg. 

I'm unsure how to go about validating the environment in which the package is being built to check the presence of the cargo related variables.

Reply:
In `default.nix` only specific parts of the source tree are being included in the build, and the `examples` directory was not being included. Instead of using the examples directory in the unit tests I am instead creating temporary file trees.